engine/src/main/java/org/pentaho/di/trans/streaming/common/BaseStreamStepMeta.java

```diff
@@ -75,20 +80,44 @@
     BeanInjectionInfo info = new BeanInjectionInfo( this.getClass() );
     BeanInjector injector = new BeanInjector( info );
     Map<String, BeanInjectionInfo.Property> properties = info.getProperties();
+
+    StringBuilder buider = new StringBuilder();
+
+    info.getGroups()
+      .forEach( group -> {
+        boolean groupHasName = !isNullOrEmpty( group.getName() );
+        if ( groupHasName ) {
+          buider.append( "<" ).append( group.getName() ).append( ">\n" );
+        }
+        List<BeanInjectionInfo.Property> groupProps = group.getGroupProperties();
+        buider.append( getXmlForProps( injector,
+
+          properties.entrySet().stream()
+            .filter( entry -> groupProps.contains( entry.getValue() ) )
+            .collect( Collectors.toMap( Map.Entry::getKey, Map.Entry::getValue ) )
+        ) );
+        if ( groupHasName ) {
+          buider.append( "</" ).append( group.getName() ).append( ">\n" );
+        }
+      } );
+    return buider.toString();
```

Comment:
I'm also wondering if a map to string and a Collectors.joining would be simpler than the StringBuilder.

Reply:
I ended up switching to a StringBuilder after needing to conditionally insert a <group> tag.  Including the opening/closing tag was getting messy, and really this whole mess should probably be done with a DocumentBuilder or real serialization library so we're not assembling string xml.  Didn't want to go that far now, though.